library/src/scala/quoted/Quotes.scala

```diff
@@ -45,22 +45,24 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
      *  Returns `None` if the expression does not represent a value or possibly contains side effects.
      *  Otherwise returns the `Some` of the value.
      */
-    def value(using FromExpr[T]): Option[T] =
-      given Quotes = Quotes.this
-      summon[FromExpr[T]].unapply(self)
+    def value(using FromExpr[T]): Option[T]
 
     /** Return the value of this expression.
      *
      *  Emits an error and throws if the expression does not represent a value or possibly contains side effects.
      *  Otherwise returns the value.
      */
-    def valueOrError(using FromExpr[T]): T =
-      val fromExpr = summon[FromExpr[T]]
-      def reportError =
-        val msg = s"Expected a known value. \n\nThe value of: ${self.show}\ncould not be extracted using $fromExpr"
-        reflect.report.throwError(msg, self)
-      given Quotes = Quotes.this
-      fromExpr.unapply(self).getOrElse(reportError)
+    // TODO: deprecate in 3.1.0 and remove @experimental from valueOrAbort
+    // @deprecated("Use valueOrThrow", "3.1.0")
+    def valueOrError(using FromExpr[T]): T
+
+    /** Return the value of this expression.
+     *
+     *  Emits an error and aborts if the expression does not represent a value or possibly contains side effects.
+     *  Otherwise returns the value.
+     */
+    @experimental
+    def valueOrAbort(using FromExpr[T]): T
 
```

Comment:
I'd suggest the method to take a message for indicating the reason why it's aborted. It helps end-users in debugging and saves the efforts of macro-authors in writing error-handling code.

Reply:
The question then is if we should keep the version with the default error message. The idea if this signature was also to make the authors not need to write the error message.

Currently the can do 
```scala
x.value match
  case Some(n) => ...
  case None => reflect.report.error("......", x)
```
or just
```scala
if x.valueOrAbort == 0 then ...
else 
```

the proposed one would need to add the message
```scala
if x.valueOrAbort(".........") == 0 then
else ...
```

Reply:
Indeed, the version with the error message does not look better in code and can be a little annoying for prototyping. For macro authors, it's not clear that it's better.

Macros can be subtle to use,  more debuggability to end-users will make them more friendly for usage. However, if the gain for end-users is small, maybe it's not worth the complication.

> The question then is if we should keep the version with the default error message

We probably only want to keep one of the two.

Reply:
We could add an `valueOrAbortWith("....")` later or users will just define their own in the meantime. We should probably wait and see what users define themselves to know what is preferable in paractice.

Reply:
Meanwhile, they can just define something like this in their projects

```scala
extension [T: FromExpr](expr: Expr[T]) 
  def valueOrAbortWith(msg: String)(using Quotes) = 
    x.value match
      case Some(n) => n
      case None => reflect.report.errorAndAbort(msg, x)
```

```scala
Expr(1).valueOrAbortWith("expected a constant")
```

Reply:
I agree, it's better to wait and do it later.